File: apps/c/blink.c
#include <commands.h>

#define REGISTER_APP(name, init) \
  void (*name)() __attribute__((section(".app.blink"))) = init

static uint32_t count = 0;

void
timer_fired() {
  if (!(++count % 10)) {
    println("That's 10 timer fires.");
  }

  timer_subscribe(1 << 15, timer_fired);
  wait();
}

void
init() {
  toggle_led();
  println("Welcome to the blink app!");

  timer_subscribe(1 << 15, timer_fired);
  wait();
}

REGISTER_APP(blink, init);

File: apps/rust/blink.rs
static mut count: usize = 0;

fn timer_fired() {
    unsafe {
        count += 1;
        if count % 7 == 0 {
            super::println("Rust: Boop.");
        }
    }

    super::timer_subscribe(1 << 15, timer_fired);
    super::wait();
}

fn initialize() {
    super::println("I'm in the Rust app!");
    super::timer_subscribe(1 << 15, timer_fired);
    super::wait();
}

#[link_section = ".app.rust-boop"]
pub static RUST_BLINK_INIT: fn() = initialize;
